src/Framework/PaymentGateways/PaymentGateway.php

```diff
@@ -57,6 +55,38 @@ public function __construct(SubscriptionModule $subscriptionModule = null)
         $this->subscriptionModule = $subscriptionModule;
     }
 
+    /**
+     * @unreleased
+     */
+    public static function isDeprecated(): bool
+    {
+        // To help with migration, gateways are considered deprecated until GiveWP 3.0 until they specify otherwise.
+        // TODO: determine if we should hardcode this as true and let gateways specify themselves until GIVE 4.0 or just assume all gateways are compatible with GIVE 3.0
+        return !(defined('GIVE_VERSION') && version_compare(GIVE_VERSION, '3.0.0', '>='));
+    }
```

Comment:
This is worth a discussion on how to handle migrating gateways to support v3 forms.  In an ideal world, all gateways will be active unless specifying they are deprecated (not compatible with v3 forms).

To be clear, our framework decides how we handle deprecated gateways and for 3.0 if a gateway is marked as deprecated it will not work on v3 forms.  Backwards compatibility for v2 forms is determined by the use of the method `getLegacyFormFieldMarkup`.

You might be thinking: _What about the presence of a gateway script?_(`enqueueScript()` method)
Well, thinking about some gateways like offsite ones that won't necessarily need a js file - we don't want to require it.  So this should not be the source of truth for v3 compatibility

The problem is there are a mix of gateways in the wild and additional add-ons that have not been updated yet.

So we have 2 options.

1. Leave this as true for the duration of 3.0 by default and let gateways mark this as false individually to declare compatibility with v3 forms.  Then by 4.0 this would be false by default.  In other words, _all_ gateways are deprecated for 3.0 until they are updated.
2. Basically do the opposite and let gateways mark themselves as deprecated explicitly.  In other words, we would be assuming all gateways are compatible with 3.0 unless explicitly marked as deprecated.


Personally, I think option 1 would be the safest bet and easiest to manage migrating gateways.  The only real downside is having to overload the `isDeprecated()` method on each gateways that will eventually be false by default.  This would also be necessary for _new_ gateways to overload for the duration of 3.0.

Reply:
@glaubersilva @kjohnson @JasonTheAdams would like to get all your thoughts on this.

Reply:
**Update**

Talked to @JasonTheAdams about this and we came to the conclusion that it would be wise to start versioning our gateway api which just happens to correspond to our form versions.  Eventually these could diverge like v4 forms but gateways on api v3 would still work.  This would mean instead of deprecating gateways, they would announce what gateway api version they support.  This will come from either a version specific interface(s) or a simple array of version numbers.

Reply:
We also want to go with option 1 which means all _existing_ gateways do not work on v3 forms until they explicitly define they do (api version)